File: Bindings/Python/tests/test_access_subcomponents.py
"""
Test that sockets, inputs, and outputs are functional in python.
"""
import os, unittest
import opensim as osim

test_dir = os.path.join(os.path.dirname(os.path.abspath(osim.__file__)),
                        'tests')

# Silence warning messages if mesh (.vtp) files cannot be found.
osim.Model.setDebugLevel(0)

class TestAccessSubcomponents(unittest.TestCase):
    def test_individual_components(self):
        model = osim.Model(os.path.join(test_dir, "arm26.osim"))
        muscle = model.getComponent('forceset/BICshort')
        assert muscle.getName() == 'BICshort'
        # No downcasting necessary!
        muscle.get_max_isometric_force() # Method on Muscle.
        muscle = model.updComponent('forceset/BICshort')
        muscle.set_max_isometric_force(100)

    def test_component_list(self):
        model = osim.Model(os.path.join(test_dir, "arm26.osim"))

        num_components = 0
        for comp in model.getComponentsList():
            num_components += 1
        assert num_components > 0

        num_bodies = 0
        for body in model.getBodyList():
            num_bodies += 1
            body.getMass()
        assert num_bodies == 2

        num_joints = 0
        for joint in model.getJointList():
            num_joints += 1
            joint.numCoordinates()
        assert num_joints == 2

        # Custom filtering.
        num_bodies = 0
        for frame in model.getFrameList():
            body = osim.Body.safeDownCast(frame)
            if body != None:
                num_bodies += 1
                print(body.getName())
                body.getInertia()
        assert num_bodies == 2

        model = osim.Model()
        ground = model.getGround()

        thelenMuscle = osim.Thelen2003Muscle("Darryl", 1, 0.5, 0.5, 0)
        thelenMuscle.addNewPathPoint("muscle1-point1", ground, osim.Vec3(0.0,0.0,0.0))
        thelenMuscle.addNewPathPoint("muscle1-point2", ground, osim.Vec3(1.0,0.0,0.0))
        millardMuscle = osim.Millard2012EquilibriumMuscle("Matt", 1, 0.5,
                                                          0.5, 0)
        millardMuscle.addNewPathPoint("muscle1-point1", ground, osim.Vec3(0.0,0.0,0.0))
        millardMuscle.addNewPathPoint("muscle1-point2", ground, osim.Vec3(1.0,0.0,0.0))

        model.addComponent(thelenMuscle)
        model.addComponent(millardMuscle)

        # Total number of muscles is 2.
        assert len(set(model.getMuscleList())) == 2
        for muscle in model.getMuscleList():
            assert (isinstance(muscle, osim.Thelen2003Muscle) or
                    isinstance(muscle, osim.Millard2012EquilibriumMuscle))

        # There is exactly 1 Thelen2003Muscle.
        assert len(set(model.getThelen2003MuscleList())) == 1
        for muscle in model.getThelen2003MuscleList():
            assert isinstance(muscle, osim.Thelen2003Muscle)

        # There is exactly 1 Millard2012EquilibriumMuscle.
        assert len(set(model.getMillard2012EquilibriumMuscleList())) == 1
        for muscle in model.getMillard2012EquilibriumMuscleList():
            assert isinstance(muscle, osim.Millard2012EquilibriumMuscle)

    def test_component_filter(self):
        model = osim.Model(os.path.join(test_dir, "arm26.osim"))
        comps = model.getMuscleList()
        comps.setFilter(osim.ComponentFilterAbsolutePathNameContainsString('BIC'))
        count = 0
        BICnames = ['BIClong', 'BICshort']
        for comp in comps:
            assert comp.getName() == BICnames[count]
            # The ComponentList iterator does the downcasting for us!
            assert type(comp) == osim.Thelen2003Muscle
            count += 1
        assert count == 2

File: Bindings/Java/Matlab/tests/testAccessSubcomponents.m

import org.opensim.modeling.*

model = Model('arm26.osim');

% Individual components
% =====================
muscle = model.getComponent('forceset/BICshort');
assert(strcmp(muscle.getName(), 'BICshort'));
thelenMuscle = Thelen2003Muscle.safeDownCast(muscle);
thelenMuscle.get_max_isometric_force();
muscle = model.updComponent('forceset/BICshort');
updThelenMuscle = Thelen2003Muscle.safeDownCast(muscle)
updThelenMuscle.set_max_isometric_force(100);

% ComponentList
% =============

% Generic.
comps = model.getComponentsList();
it = comps.begin();
countComps = 0;
while ~it.equals(comps.end())
    it.getName(); % Object.
    it.getNumSockets(); % Component.
    it.next();
    countComps = countComps + 1;
end
assert(countComps > 0);

% Specific type of component.
muscleList = model.getMuscleList();
muscleIter = muscleList.begin();
countMuscles = 0;
while ~muscleIter.equals(muscleList.end())
    % SWIG puts all of Muscle's methods on MuscleIterator.
    muscleIter.get_max_isometric_force();
    % Access the underlying Muscle.
    % TODO MATLAB methods can't start with underscore muscle = muscleIter.__deref__();
    % TODO MATLAB methods can't start with underscore muscle.get_max_isometric_force();
    muscleIter.next();
    countMuscles = countMuscles + 1;
end
assert(countMuscles > 0);

% Custom type filter.
% TODO does not work (yet).
% TODO it = comps.begin();
% TODO while ~it.equals(comps.end())
% TODO     disp(it.getConcreteClassName());
% TODO     if it.isA('Thelen2003Muscle')
% TODO         disp(it.getConcreteClassName());
% TODO     end
% TODO     it.next();
% TODO end

% Use ComponentList filter.
muscleList.setFilter(ComponentFilterAbsolutePathNameContainsString('BIC'));
muscleIter = muscleList.begin();
countBICMuscles = 0;
BICnames = {'BIClong', 'BICshort'};
while ~muscleIter.equals(muscleList.end())
    countBICMuscles = countBICMuscles + 1;
    disp(muscleIter.getName());
    assert(strcmp(BICnames{countBICMuscles}, muscleIter.getName()));
    muscleIter.next();
end





